src/Phan/Language/Type/ArrayShapeType.php

```diff
@@ -527,7 +527,9 @@ public static function fromFieldTypes(
             $key_parts[$key] = $field_union_type->generateUniqueId();
         }
         if ($is_nullable) {
-            $key_parts[] = '?';
+            $key_parts = [ '?' => $key_parts ];
+        } else {
+            $key_parts = [ ' ' => $key_parts ];
         }
         $key = \json_encode($key_parts);
```

Comment:
This would be wrong if the array contained `'?'` or `' '` as keys.

I just need an unambiguous identifier - appending `?` after JSON (or prepending) is unambiguous.

```suggestion
        $key = \json_encode($key_parts) . ($is_nullable ? '?' : '');
```

Reply:
It's not wrong, because we're adding a new level to the array, not adding a new key.  (Adding a new key is how the original code went wrong.)  That said, your solution is perfectly fine too if you prefer it.

Reply:
> It's not wrong, because we're adding a new level to the array, not adding a new key. (Adding a new key is how the original code went wrong.) That said, your solution is perfectly fine too if you prefer it.

Oh, you're right, I misread that.